components/wmtk_components/marching/internal/Marching.cpp

```diff
@@ -0,0 +1,112 @@
+    case 3: process_in_3d(); break;
+    default: throw std::runtime_error("settings went wrong!"); break;
+    }
+}
+
+void Marching::process_in_2d()
+{
+    using namespace operations;
+
+    wmtk::MeshAttributeHandle<long> todo_edgesplit_same_handle = m_mesh.register_attribute<long>(
+        std::string("todo_edgesplit_same_tag"),
+        wmtk::PrimitiveType::Edge,
+        1);
+
+    wmtk::Accessor<long> acc_vertex_tag = m_mesh.create_accessor(m_vertex_tag);
+    wmtk::Accessor<long> acc_edge_tag = m_mesh.create_accessor(m_edge_tag);
+    wmtk::Accessor<double> acc_pos = m_mesh.create_accessor(m_position_handle);
+    wmtk::Accessor<long> acc_todo_edgesplit_same_tag =
+        m_mesh.create_accessor(todo_edgesplit_same_handle);
+
+    // edge split
+    {
+        // compute the todo list for the split edge with the same ends
+        const std::vector<Tuple>& edges = m_mesh.get_all(wmtk::PrimitiveType::Edge);
+        for (const Tuple& edge : edges) {
+            long vt0, vt1, et0;
+            vt0 = acc_vertex_tag.scalar_attribute(edge);
+            vt1 = acc_vertex_tag.scalar_attribute(m_mesh.switch_vertex(edge));
+            if ((vt0 == m_input_tag_value && vt1 == m_embedding_tag_value) ||
+                (vt1 == m_input_tag_value && vt0 == m_embedding_tag_value)) {
+                acc_todo_edgesplit_same_tag.scalar_attribute(edge) = 1;
+            }
+        }
+        // using scheduler to do edge splitting
+        OperationSettings<tri_mesh::EdgeSplitWithTag> settings_split;
+        settings_split.edge_tag = m_edge_tag;
+        settings_split.vertex_tag = m_vertex_tag;
+        settings_split.embedding_tag_value = m_embedding_tag_value;
+        settings_split.need_embedding_tag_value = true;
+        // settings_split_same.position = m_position_handle;
+        settings_split.split_with_tag_settings.split_settings.split_boundary_edges = true;
+        settings_split.split_with_tag_settings.position = m_position_handle;
+        settings_split.split_edge_tag_value = m_embedding_tag_value;
+        settings_split.split_vertex_tag_value = m_split_tag_value;
+        settings_split.split_todo = todo_edgesplit_same_handle;
+        settings_split.initialize_invariants(m_mesh);
+        m_scheduler.add_operation_type<tri_mesh::EdgeSplitWithTag>("edge_split", settings_split);
+        while (true) {
+            m_scheduler.run_operation_on_all(PrimitiveType::Edge, "edge_split");
+            if (m_scheduler.number_of_successful_operations() == 0) {
+                break;
+            }
+        }
+    }
+
+    // link the split vertices
+    {
+        for (const Tuple& t : m_mesh.get_all(PrimitiveType::Edge)) {
+            long vt0, vt1;
+            vt0 = acc_vertex_tag.const_scalar_attribute(t);
+            vt1 = acc_vertex_tag.const_scalar_attribute(m_mesh.switch_vertex(t));
+            if (vt0 == m_split_tag_value && vt1 == m_split_tag_value) {
+                acc_edge_tag.scalar_attribute(t) = m_split_tag_value;
```

Comment:
This code sounds to me like it should not be inside this component. It is very specific to your project.